### PR TITLE
feat: add ST_Relate(geometry, geometry, text) boolean variant

### PR DESCRIPTION
**Pull request by Mehak3010**
Adds the boolean overload of ST_Relate:

boolean ST_Relate(geometry geomA, geometry geomB, text intersectionMatrixPattern)

This allows checking if two geometries satisfy a given DE-9IM pattern, complementing the existing text-returning variant.

- Uses GEOS relate_pattern
- Includes tests for pattern matching and null handling
This work addresses issue #539 and is part of the broader implementation tracked in ossue #174 .

### CODE REVIEW

**Comment by Mehak3010** [comment]
@petern48 CI checks are all passing now. The PR #174  is ready for review — let me know if any changes are needed.

**Review by petern48 — COMMENTED** [review]
Looks like a good start. There are still a few things missing:

- Add Python integration tests. You can start by using the same test cases as the Python tests added by your previous [st_relate PR](https://github.com/apache/sedona-db/pull/691) (except inverted) + including some cases that evaluate to `False`.

- Update the docs (`docs/reference/sql/st_relate.qmd`) with this new syntax. You can take a look at [st_buffer.qmd](https://github.com/apache/sedona-db/blob/0163a6bece98a127c44a74d4a9c9bb5e42f5f3d9/docs/reference/sql/st_buffer.qmd) for an example of how to write docs for a function with multiple kernels.

- Minor nit. Could you link the issue in your PR description? You can just put something like "part of #ISSUE."

At a quick skim, things look good so far. I'll take a more thorough look once these are in place.

**Comment by Mehak3010** [comment]
@petern48 Thanks for the feedback! I’ve added the Python integration tests (including false cases), updated the docs with the new syntax, and linked the issue in the PR description.
All CI checks are passing now — ready for further review.

File: python/sedonadb/tests/functions/test_predicates.py [hunk 1]
```diff
@@ -495,3 +495,55 @@ def test_st_relate(eng, geom1, geom2, expected):
...
+            "212101212",
+            True,
+        ),
+    ],
+)
+def test_st_relate_pattern(eng, geom1, geom2, pattern, expected):
+    eng = eng.create_or_skip()
+    pattern_sql = "NULL" if pattern is None else f"'{pattern}'"
+    eng.assert_query_result(
+        f"SELECT ST_Relate({geom_or_null(geom1)}, {geom_or_null(geom2)}, {pattern_sql})",
```
Comment (petern48) [inline python/sedonadb/tests/functions/test_predicates.py:582]:
```suggestion
    eng.assert_query_result(
        f"SELECT ST_Relate({geom_or_null(geom1)}, {geom_or_null(geom2)}, '{val_or_null(pattern)}')",
```

nit: We have an [existing helper](https://github.com/Mehak3010/sedona-db/blob/2822b30680a7b46ddd12fd1f32eacb576a16abac/python/sedonadb/python/sedonadb/testing.py#L678-L692) for this :)


edit: this suggestion was wrong. it's not the exact same logic. see https://github.com/apache/sedona-db/pull/741#discussion_r3036161388. Sorry for misleading you.

File: python/sedonadb/tests/functions/test_predicates.py [hunk 2]
```diff
@@ -495,3 +495,55 @@ def test_st_relate(eng, geom1, geom2, expected):
...
     )
+
+
+@pytest.mark.parametrize("eng", [SedonaDB, PostGIS])
+@pytest.mark.parametrize(
+    ("geom1", "geom2", "pattern", "expected"),
+    [
+        (None, None, None, None),
+        ("POINT (0 0)", None, "FF0FFF0F2", None),
+        (None, "POINT (0 0)", "FF0FFF0F2", None),
```
Comment (petern48) [inline python/sedonadb/tests/functions/test_predicates.py:506]:
```suggestion
        (None, "POINT (0 0)", "FF0FFF0F2", None),
        ("POINT (0 0)", "POINT (0 0)", None, None),
```
nice to test this variation too (only pattern is `NULL`)

File: python/sedonadb/tests/functions/test_predicates.py
Comment (petern48) [inline python/sedonadb/tests/functions/test_predicates.py:489]: Could we add a few more of these same test cases from `test_st_relate` as `True` cases for `test_st_relate_pattern`? e.g., we're missing a point-in-polygon-hole case and linestring and geometry collections cases, a few lines up.

Maybe add a few more `False` cases for other predicates, since we only have two tests that return `False` at the moment.

**Review by petern48 — COMMENTED** [review]
Looking great. I have a bit more feedback about testing, but I think this should be ready afterwards.

File: python/sedonadb/tests/functions/test_predicates.py
Comment (Mehak3010) [reply python/sedonadb/tests/functions/test_predicates.py:489]: yeah sure! working on it.

**Comment by Mehak3010** [comment]
@petern48 Added the suggested cases — included point-in-polygon-hole, linestring, and geometry collection cases, along with additional False cases for better coverage.
Let me know if anything else should be added!

File: python/sedonadb/tests/functions/test_predicates.py (hunk 2 above)
Comment (petern48) [reply python/sedonadb/tests/functions/test_predicates.py:506]: I see that you undid this after it failed

After looking at the [CI failure](https://github.com/apache/sedona-db/actions/runs/23986451946/job/69959059369), I realized we're passing in `'NULL'` instead of `NULL` into the function.

```
E   adbc_driver_manager.ProgrammingError: INVALID_ARGUMENT: [libpq] Failed to execute query: could not begin COPY: ERROR:  GEOSRelatePattern: IllegalArgumentException: Unknown dimension symbol: N
E   
E   
E   Query was: COPY (SELECT ST_Relate(ST_GeomFromText('POINT (0 0)'), ST_GeomFromText('POINT (0 0)'), 'NULL')) TO STDOUT (FORMAT binary). SQLSTATE: XX000
```

This is a result of applying my suggestion here https://github.com/apache/sedona-db/pull/741#discussion_r3035740312. That's my bad. Let's revert that change (https://github.com/apache/sedona-db/pull/741/changes/78914a759fd080b18cd38377e826d6873b54aee5) to avoid passing a string instead of null and then it should work. i.e.

And then add this test case back, and I expect it to work.
```python
("POINT (0 0)", "POINT (0 0)", None, None),
```

File: python/sedonadb/tests/functions/test_predicates.py [hunk 3]
```diff
@@ -156,7 +156,6 @@ def test_st_disjoint(eng, geom1, geom2, expected):
     [
         (None, "POINT (0 0)", 1, None),
         ("POINT (1 1)", None, 1.0, None),
-        ("POINT (0 0)", "POINT (0 0)", None, None),
```
Comment (petern48) [inline python/sedonadb/tests/functions/test_predicates.py:159]: Looks like you also deleted this test from `test_st_relate` (which already passes from your prev PR). Let's add this back. Should still pass since there were no changes to st_relate code.

**Review by petern48 — COMMENTED** [review]
Looks like I made a small mistake in my previous suggestion which led to those test failures. Let's bring those test cases back, but I explained how to get them to pass.

**Comment by Mehak3010** [comment]
@petern48 Thanks for the clarification! I’ve reverted the NULL handling change (now passing actual `NULL`), re-added the removed test case, and kept all suggested edge cases. Everything should now align with the expected behavior — ready for review.

**Review by petern48 — APPROVED** [review]

File: python/sedonadb/tests/functions/test_predicates.py (hunk 2 above)
Comment (petern48) [reply python/sedonadb/tests/functions/test_predicates.py:506]: Actually, https://github.com/apache/sedona-db/pull/741/changes/516ce5f68e78ed47220167a254afc86e4fa2aa63 only restored the test to `test_st_relate`, which does address https://github.com/apache/sedona-db/pull/741#discussion_r3036168644, but does not address this current comment. I requested to restore both. Could you please restore it to `test_st_relate_pattern` as well? This is important because we technically still haven't seen proof that this is working.

**Review by petern48 — CHANGES_REQUESTED** [review]
See https://github.com/apache/sedona-db/pull/741#discussion_r3041158662

**Review by Mehak3010 — COMMENTED** [review]
Thanks for the detailed review and helpful suggestions! I’ve addressed all the changes across tests, docs, and implementation.

**Review by petern48 — APPROVED** [review]
Thanks!

**Comment by Mehak3010** [comment]
Thanks @petern48, really appreciate the support and guidance throughout!

Also, building on this work, I’m exploring a proposal around extending boolean spatial predicates and improving API consistency in Sedona. Would love to hear any suggestions or directions if you have in mind. @paleolimbot, @petern48

**Comment by petern48** [comment]
> I’m exploring a proposal around extending boolean spatial predicates and improving API consistency in Sedona.

Sorry, I don't quite follow. Do you mean adding new predicates or improving existing ones? From the knowledge, I don't know of any further improvements we could make to existing boolean predicates (they already follow proper API consistency). I also don't know of any more predicates to newly introduce.

**Comment by Mehak3010** [comment]
Thanks for the clarification!

I’ll take a closer look and try to identify concrete examples (if any) where behavior could be improved or made more consistent across SQL and Python.

If everything is already aligned as expected, I’ll focus on understanding the current design better and look into other areas where boolean predicate usability or coverage could be extended.
